sentry/src/main/java/io/sentry/SentryUUID.java

```diff
@@ -0,0 +1,299 @@
+    uuidChars[0] = HEX_DIGITS[(int) ((mostSignificantBits & 0xf000000000000000L) >>> 60)];
+    uuidChars[1] = HEX_DIGITS[(int) ((mostSignificantBits & 0x0f00000000000000L) >>> 56)];
+    uuidChars[2] = HEX_DIGITS[(int) ((mostSignificantBits & 0x00f0000000000000L) >>> 52)];
+    uuidChars[3] = HEX_DIGITS[(int) ((mostSignificantBits & 0x000f000000000000L) >>> 48)];
+    uuidChars[4] = HEX_DIGITS[(int) ((mostSignificantBits & 0x0000f00000000000L) >>> 44)];
+    uuidChars[5] = HEX_DIGITS[(int) ((mostSignificantBits & 0x00000f0000000000L) >>> 40)];
+    uuidChars[6] = HEX_DIGITS[(int) ((mostSignificantBits & 0x000000f000000000L) >>> 36)];
+    uuidChars[7] = HEX_DIGITS[(int) ((mostSignificantBits & 0x0000000f00000000L) >>> 32)];
+    uuidChars[8] = HEX_DIGITS[(int) ((mostSignificantBits & 0x00000000f0000000L) >>> 28)];
+    uuidChars[9] = HEX_DIGITS[(int) ((mostSignificantBits & 0x000000000f000000L) >>> 24)];
+    uuidChars[10] = HEX_DIGITS[(int) ((mostSignificantBits & 0x0000000000f00000L) >>> 20)];
+    uuidChars[11] = HEX_DIGITS[(int) ((mostSignificantBits & 0x00000000000f0000L) >>> 16)];
+    uuidChars[12] = HEX_DIGITS[(int) ((mostSignificantBits & 0x000000000000f000L) >>> 12)];
+    uuidChars[13] = HEX_DIGITS[(int) ((mostSignificantBits & 0x0000000000000f00L) >>> 8)];
+    uuidChars[14] = HEX_DIGITS[(int) ((mostSignificantBits & 0x00000000000000f0L) >>> 4)];
+    uuidChars[15] = HEX_DIGITS[(int) (mostSignificantBits & 0x000000000000000fL)];
+  }
+
+  static long getHexValueForChar(final char c) {
+    try {
+      if (HEX_VALUES[c] < 0) {
+        throw new IllegalArgumentException("Illegal hexadecimal digit: " + c);
+      }
+    } catch (final ArrayIndexOutOfBoundsException e) {
+      throw new IllegalArgumentException("Illegal hexadecimal digit: " + c);
+    }
+
+    return HEX_VALUES[c];
+  }
+
+  public static long randomHalfLengthUUID() {
+    Random random = SentryUUID.Holder.numberGenerator;
+    byte[] randomBytes = new byte[8];
+    random.nextBytes(randomBytes);
+    randomBytes[6] &= 0x0f; /* clear version        */
+    randomBytes[6] |= 0x40; /* set to version 4     */
+
+    long msb = 0;
+
+    for (int i = 0; i < 8; i++) msb = (msb << 8) | (randomBytes[i] & 0xff);
+
+    return msb;
+  }
+
+  public static UUID randomUUID() {
+    Random random = SentryUUID.Holder.numberGenerator;
+    byte[] randomBytes = new byte[16];
+    random.nextBytes(randomBytes);
+    randomBytes[6] &= 0x0f; /* clear version        */
+    randomBytes[6] |= 0x40; /* set to version 4     */
+    randomBytes[8] &= 0x3f; /* clear variant        */
+    randomBytes[8] |= (byte) 0x80; /* set to IETF variant  */
+
+    long msb = 0;
+    long lsb = 0;
+
+    for (int i = 0; i < 8; i++) msb = (msb << 8) | (randomBytes[i] & 0xff);
+
+    for (int i = 8; i < 16; i++) lsb = (lsb << 8) | (randomBytes[i] & 0xff);
+
+    return new UUID(msb, lsb);
+  }
+
+  private static class Holder {
+    static final Random numberGenerator = new Random();
```

Comment:
According to https://docs.oracle.com/javase/7/docs/api/java/util/Random.html there might be poor performance when using `Random` from multiple threads.

> Instances of java.util.Random are threadsafe. However, the concurrent use of the same java.util.Random instance across threads may encounter contention and consequent poor performance. Consider instead using [ThreadLocalRandom](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadLocalRandom.html) in multithreaded designs.

Should we vendor `ThreadLocalRandom` as well and use that?

Did you also do (performance) testing in a multi threaded environment?

Reply:
Hmm, good point. No, didn't do multithreaded perf testing. could try to spin up something to compare `Random` to `ThreadLocalRandom`

Reply:
Yes, this does make quite a difference. Below is a comparison for UUID generation:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<!--
<style>
table
	{mso-displayed-decimal-separator:"\,";
	mso-displayed-thousand-separator:"\.";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl63
	{mso-number-format:"0\.000";}
.xl64
	{mso-number-format:0%;}

</style>
-->
</head>

<body link="#467886" vlink="#96607D">


Benchmark | Score |   | Error | Units | Threads | Comparison
-- | -- | -- | -- | -- | -- | --
uuidGenerationRandom | 23789834,814 | ± | 117697,18 | ops/s | 1 | 100%
uuidGenerationRandomMultiThreaded | 2280766,845 | ± | 263522,36 | ops/s | 10 | 10%
uuidGenerationThreadLocalRandomMultiThreaded | 185639075,638 | ± | 5090316,9 | ops/s | 10 | 780%



</body>

</html>

Reply:
whew, what's the baseline we're comparing to here? are all of those using the FastUUID code just with a different `Random` implementation?

Reply:
The code from `FastUUID` is only responsible for converting the `UUID` to string. This was not done in this test.

I tested the `SentryUUID.randomUUID()` function, which is a clone of the `java.util.UUID.randomUUID()` function with `java.util.Random` as the random generator.

So, Results 1 and 2 test the above mentioned function in single-threaded vs multi-threaded scenarios.

Result 3 tests the same function with `java.util.Random` replaced with `java.util.concurrent.ThreadLocalRandom`. I replaced `Random random = SentryUUID.Holder.numberGenerator;` with `Random random = ThreadLocalRandom.current();`

Reply:
Also, the benchmark code can now be found here: https://github.com/lbloder/senty-uuid-bench